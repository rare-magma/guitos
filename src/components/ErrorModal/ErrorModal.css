.accordion-button:after,
.accordion-button:not(.collapsed):after {
  filter: invert(58%) sepia(93%) saturate(2284%) hue-rotate(322deg)
    brightness(95%) contrast(112%);
}

.accordion-button,
.accordion-button:after,
.accordion-button:not(.collapsed):after,
.accordion-flush,
.accordion-item,
.modal-item {
  border-radius: 0.375rem;
}

.accordion-button:focus,
.accordion-item:focus {
  border-radius: 0.375rem;
  box-shadow: 0 0 0 0.2rem rgba(189, 147, 249, 0.25);
}

.accordion,
.accordion-button,
.accordion-button:focus,
.accordion-item,
.accordion-item:focus {
  background: var(--lightbgcolor);
  border: 0 solid var(--comment);
  border-radius: 0.375rem;
  color: var(--textcolor);
}

.btn-delete-modal {
  background-color: var(--lightbgcolor);
  border: 1px solid var(--comment);
  border-radius: 0.375rem;
  color: var(--red);
}

.btn-delete-modal:focus-visible,
.btn-delete-modal:hover {
  background-color: var(--red);
  border: 1px solid var(--red);
  border-radius: 0.375rem;
  color: var(--lightbgcolor);
}

.modal-90w {
  max-width: none !important;
  width: 90%;
}

.modal-body {
  max-height: 65vh;
  overflow: auto;
}

.modal-content,
.modal-item {
  background: var(--lightbgcolor);
  border: 1px solid var(--comment);
  color: var(--textcolor);
}

.modal-header {
  border-bottom: 0;
}
