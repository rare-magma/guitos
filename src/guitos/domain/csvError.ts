import type { ParseError } from "papaparse";

export class CsvError {
  errors: ParseError[];
  file: string;

  constructor(errors: ParseError[], file: string) {
    this.errors = errors;
    this.file = file;
  }
}
