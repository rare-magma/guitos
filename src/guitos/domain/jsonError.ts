export class JsonError {
  errors: string;
  file: string;

  constructor(errors: string, file: string) {
    this.errors = errors;
    this.file = file;
  }
}
