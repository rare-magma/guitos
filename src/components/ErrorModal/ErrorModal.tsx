import { ParseError } from "papaparse";
import { Accordion, Button, Modal } from "react-bootstrap";
import { BsXLg } from "react-icons/bs";
import "./ErrorModal.css";

export interface CsvError {
  errors: ParseError[];
  file: string;
}

export interface JsonError {
  errors: string;
  file: string;
}

interface ErrorModalProps {
  error: string | null;
  show: boolean;
  jsonError: JsonError[];
  csvError: CsvError[];
  onShow: (value: boolean) => void;
  onError: () => void;
}

function ErrorModal({
  error,
  show,
  jsonError,
  csvError,
  onError,
  onShow,
}: ErrorModalProps) {
  const showModal = error && show;
  const showJsonError = jsonError && jsonError.length > 0 && show;
  const showCsvError = csvError && csvError.length > 0 && show;

  function handleShow(value: boolean) {
    onShow(value);
  }

  function handleError() {
    onError();
  }

  return (
    <>
      {showModal && (
        <Modal
          data-testid="error-modal"
          dialogClassName="modal-90w mx-auto"
          show={show}
          onHide={() => handleShow(false)}
          centered
        >
          <Modal.Header>
            Error:
            <Button
              className="align-self-end"
              data-testid="error-modal-dismiss"
              key={"modal-dismiss-button"}
              variant="delete-modal"
              type="button"
              onClick={() => {
                handleShow(false);
              }}
            >
              <BsXLg aria-hidden />
            </Button>
          </Modal.Header>
          <Modal.Body className="textarea mx-1">
            <p className="code">{error}</p>
          </Modal.Body>
        </Modal>
      )}

      {showJsonError && (
        <Modal
          key="json-error-modal"
          data-testid="json-error-modal"
          dialogClassName="modal-90w mx-auto"
          show={show}
          onHide={() => handleShow(false)}
          centered
        >
          <Modal.Header key="json-error-modal-header">
            Errors found while importing:
            <Button
              data-testid="json-error-close"
              className="align-self-end"
              aria-label="close error dialog"
              key={"modal-dismiss-button"}
              variant="delete-modal"
              type="button"
              onClick={() => {
                handleShow(false);
                handleError();
              }}
            >
              <BsXLg aria-hidden />
            </Button>
          </Modal.Header>
          <Modal.Body key="json-error-modal-body">
            <Accordion
              aria-label="error list"
              key="json-error-modal-accordion"
              flush
            >
              {jsonError.map((jsonError: JsonError, i: number) => (
                <Accordion.Item
                  key={`${jsonError.file}-item-${i}`}
                  eventKey={jsonError.file}
                >
                  <Accordion.Header
                    aria-label="file"
                    key={`${jsonError.file}-header-${i}`}
                  >
                    {jsonError.file}
                  </Accordion.Header>
                  <Accordion.Body
                    className="textarea fixed-width-font mx-1"
                    key={`${jsonError.file}-body-${i}`}
                  >
                    <p className="code" key={`${jsonError.file}-error-${i}`}>
                      <>
                        {jsonError.errors}
                        <br />
                      </>
                    </p>
                  </Accordion.Body>
                </Accordion.Item>
              ))}
            </Accordion>
          </Modal.Body>
        </Modal>
      )}

      {showCsvError && (
        <Modal
          key="csv-error-modal"
          data-testid="csv-error-modal"
          dialogClassName="modal-90w mx-auto"
          show={show}
          onHide={() => handleShow(false)}
          centered
        >
          <Modal.Header key="csv-error-modal-header">
            Errors found while importing:
            <Button
              data-testid="csv-error-close"
              className="align-self-end"
              key={"modal-dismiss-button"}
              aria-label="close error dialog"
              variant="delete-modal"
              type="button"
              onClick={() => {
                handleShow(false);
                handleError();
              }}
            >
              <BsXLg aria-hidden />
            </Button>
          </Modal.Header>
          <Modal.Body key="csv-error-modal-body">
            <Accordion
              aria-label="error list"
              key="csv-error-modal-accordion"
              flush
            >
              {csvError.map((csvError: CsvError, i: number) => (
                <Accordion.Item
                  key={`${csvError.file}-item-${i}`}
                  eventKey={csvError.file}
                >
                  <Accordion.Header
                    aria-label="file"
                    key={`${csvError.file}-header-${i}`}
                  >
                    {csvError.file}
                  </Accordion.Header>
                  <Accordion.Body
                    className="textarea fixed-width-font mx-1"
                    key={`${csvError.file}-body-${i}`}
                  >
                    <p className="code" key={`${csvError.file}-csverror-${i}`}>
                      {csvError.errors.map((error) => (
                        <span key={error.row}>
                          Line {error.row}: {error.message}
                          <br />
                        </span>
                      ))}
                    </p>
                  </Accordion.Body>
                </Accordion.Item>
              ))}
            </Accordion>
          </Modal.Body>
        </Modal>
      )}
    </>
  );
}

export default ErrorModal;
