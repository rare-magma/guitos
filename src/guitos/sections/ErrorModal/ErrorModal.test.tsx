import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { describe, expect, it, vi } from "vitest";
import type { CsvError } from "../../domain/csvError";
import { CsvErrorMother } from "../../domain/csvError.mother";
import type { JsonError } from "../../domain/jsonError";
import { JsonErrorMother } from "../../domain/jsonError.mother";
import { ErrorModal } from "./ErrorModal";

const error = "Thrown error";
const jsonErrors: JsonError[] = [JsonErrorMother.error()];

const csvErrors: CsvError[] = [CsvErrorMother.error()];

const setShowError = vi.fn();
const setShowCsvErrors = vi.fn();
const setShowJsonErrors = vi.fn();
const handleDismiss = vi.fn();

describe("ErrorModal", () => {
  const comp = (
    <ErrorModal
      error={error}
      setShowError={setShowError}
      showError={true}
      jsonErrors={jsonErrors}
      setJsonErrors={setShowJsonErrors}
      csvErrors={csvErrors}
      setCsvErrors={setShowCsvErrors}
      handleDismiss={handleDismiss}
    />
  );

  it("matches snapshot", () => {
    render(comp);
    expect(comp).toMatchSnapshot();
  });

  it("renders initial state", () => {
    render(comp);
    expect(
      screen.getAllByText("Errors found while importing:")[0],
    ).toBeInTheDocument();
    expect(
      screen.getByText(
        "SyntaxError: Expected ',' or '}' after property value in JSON at position 209",
      ),
    ).toBeInTheDocument();
  });

  it("closes error when clicking the button", async () => {
    render(comp);
    expect(screen.getByTestId("error-modal")).toBeInTheDocument();
    await userEvent.click(screen.getByTestId("error-modal-dismiss"));
  });

  it("closes json error when clicking the button", async () => {
    render(comp);
    expect(screen.getByTestId("json-error-close")).toBeInTheDocument();
    await userEvent.click(screen.getByTestId("json-error-close"));
  });

  it("closes json error modal when clicking the button", async () => {
    render(comp);
    expect(screen.getByTestId("json-error-modal")).toBeInTheDocument();
    await userEvent.click(screen.getByTestId("json-error-modal"));
  });

  it("closes csv error when clicking the button", async () => {
    render(comp);

    expect(screen.getByTestId("csv-error-close")).toBeInTheDocument();
    await userEvent.click(screen.getByTestId("csv-error-close"));
  });

  it("closes csv error modal when clicking the button", async () => {
    render(comp);

    expect(screen.getByTestId("csv-error-modal")).toBeInTheDocument();
    await userEvent.click(screen.getByTestId("csv-error-modal"));
  });
});
