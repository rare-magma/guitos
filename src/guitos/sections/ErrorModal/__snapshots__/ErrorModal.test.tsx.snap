// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ErrorModal > matches snapshot 1`] = `
<ErrorModal
  csvErrors={
    [
      {
        "errors": [
          {
            "code": "TooFewFields",
            "message": "Line 0: Too few fields: expected 3 fields but parsed 2",
            "row": 0,
            "type": "FieldMismatch",
          },
        ],
        "file": "123.csv",
      },
    ]
  }
  error="Thrown error"
  handleDismiss={[MockFunction spy]}
  jsonErrors={
    [
      {
        "errors": "SyntaxError: Expected ',' or '}' after property value in JSON at position 209",
        "file": "123.json",
      },
    ]
  }
  setCsvErrors={[MockFunction spy]}
  setJsonErrors={[MockFunction spy]}
  setShowError={[MockFunction spy]}
  showError={true}
/>
`;
